### ISSUE TITLE
test(inference): add 15 edge-case unit tests for Sampler combined filtering and boundary conditions — COMMITTED

### ISSUE THREAD

**Issue by devswha** [devswha/zipcode#178]
## Summary

Added 15 edge-case unit tests to `crates/inference/src/sampler.rs` covering previously untested gaps in the token sampler's combined filtering pipeline and boundary conditions.

**Sampler test count:** 21 → 36
**Total workspace tests:** 1362 → 1377 (all pass)

## Tests Added

| # | Test | What it covers |
|---|------|----------------|
| 1 | `test_combined_top_k_and_top_p` | top_k filters first, then top_p narrows further |
| 2 | `test_repeat_penalty_with_temperature_and_top_k` | repeat penalty + temperature + top_k interaction |
| 3 | `test_repeat_penalty_on_zero_logit` | Boundary: logit == 0.0 goes to multiply branch (no-op) |
| 4 | `test_repeat_penalty_duplicate_tokens_in_history` | Same token penalized once per occurrence |
| 5 | `test_top_p_exact_single_token_probability` | top_p at exact softmax probability boundary |
| 6 | `test_large_vocabulary_sampling` | 10k-token vocab stability (no panic/overflow) |
| 7 | `test_top_k_with_ties_at_boundary` | Tied probabilities at top_k cutoff all survive |
| 8 | `test_repeat_penalty_with_empty_past_tokens` | Empty history = no penalty applied |
| 9 | `test_temperature_very_small_positive` | temp=0.0001 uses sampling path (not greedy shortcut) |
| 10 | `test_top_p_with_uniform_distribution` | Identical logits with top_p=0.5 |
| 11 | `test_softmax_all_very_negative_logits` | Logits like [-1000, -999, -998] — no underflow |
| 12 | `test_combined_repeat_penalty_top_k_top_p` | Full pipeline: penalty + top_k + top_p + temperature |
| 13 | `test_repeat_last_n_zero_means_no_penalty` | Window size 0 = empty slice = no penalty |
| 14 | `test_sampler_new_preserves_config` | Behavioral verification of config preservation |
| 15 | `test_sequential_samples_produce_valid_tokens` | 20 sequential samples all return valid indices |

## Interesting findings during testing

1. **`test_softmax_all_very_negative_logits`** — Initially expected the "least negative" logit (-999) to win with near-greedy temperature. But temperature scaling (`/0.001`) magnifies differences so much that the actual highest logit (-998) dominates. This is correct behavior but non-obvious — the test documents this numerical stability characteristic.

2. **`test_repeat_penalty_on_zero_logit`** — Confirms that `logit == 0.0` hits the multiply branch (`0.0 * penalty = 0.0`), making repeat penalty a no-op for zero-valued logits. This is by design but was previously untested.

## Commit

```
1f64b94 test(inference): add 15 edge-case unit tests for Sampler combined filtering and boundary conditions
```

## Verification

```bash
cargo test -p zipcode-inference --lib -- sampler    # 36 passed
cargo test --workspace                               # 1377 passed
cargo clippy --workspace --all-targets -- -D warnings # clean
cargo fmt --all --check                              # clean
```

## Labels

`test`, `code-quality`

**Comment by devswha** [comment]
Closing as completed on current origin/main.

Evidence:
- current origin/main / local HEAD: 5c80f7717d5b1e5c42b1e32d4047b9fd61e70c45 (ahead/behind 0 0)
- completed commit: 1f64b94 — test(inference): add 15 edge-case unit tests for Sampler combined filtering and boundary conditions
- reachability: git merge-base --is-ancestor 1f64b94 origin/main passed
- targeted verification: cargo test -p zipcode-inference --lib sampler  # 36 passed
- broad baseline note: cargo test --workspace currently has separate smoke PTY/fullscreen failures now tracked as #181, so this closure uses issue-specific targeted evidence rather than claiming a green full workspace run.
